sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -469,6 +439,39 @@ object StateStoreProvider {
   }
 }
 
+/**
+ * This is an optional trait to be implemented by [[StateStoreProvider]]s that can read fine
+ * grained state data which is replayed from a specific snapshot version. It is used by the
+ * snapshotStartBatchId option in state data source.
+ */
+trait FineGrainedStateSource {
+  /**
+   * Used by snapshotStartBatchId option when reading state generated by join operation as data
+   * source.
+   * Return an instance of [[StateStore]] representing state data of the given version.
+   * The State Store will be constructed from the batch at startVersion, and applying delta files
+   * up to the endVersion. If there is no snapshot file of batch startVersion, an exception will
+   * be thrown.
+   *
+   * @param startVersion checkpoint version of the snapshot to start with
+   * @param endVersion   checkpoint version to end with
+   */
+  def replayStoreFromSnapshot(startVersion: Long, endVersion: Long): StateStore
```

Comment:
This is a compromise to support join state. Join state is read using `SymmetricHashJoinStateManager` and it will always return `StateStore`.
https://github.com/apache/spark/blob/7f5f96cad22464e02679ab1a1c6eb08b9da039ef/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala#L479